### ISSUE TITLE
Unicode drawText "Right to Left" Language issue

### ISSUE THREAD

**Issue by aqibmumtaz** [aheckmann/gm#427]
Hi Team,

Im trying to drawText on image for "Urdu/Arabic Language" this text "السلام عليكم", text is being drawn on output image but it is being drawn "Left to Right" while it must be "Right to Left" as Arabic/Urdu are "Right to Left" languages. What m i doing wrong? How can I fix it?
 

                                                    gm("/file.png")
                                                        .fill('#FF0066')
                                                        .drawText(100, -150, "السلام عليكم", 'Center')
                                                        .font("/alvi_Nastaleeq.ttf")
                                                        .fontSize('38px')
                                                        .encoding('Unicode')

![screen shot 2015-07-31 at 11 35 37 pm](https://cloud.githubusercontent.com/assets/1073190/9014640/148ce06e-37dd-11e5-9bb2-582af87b4dd1.png)

Thanks,

**Comment by aqibmumtaz** [comment]
Any suggestion or pointer plz?

**Comment by DanMMX** [comment]
What about using `flop`? Give a look at the [example](https://github.com/aheckmann/gm/blob/master/examples/flop.js)

**Comment by aqibmumtaz** [comment]
I tried reversing and showing, but still dont work. check out this question I updated it for revering text http://stackoverflow.com/questions/31753132/unicode-drawtext-right-to-left-language-issue-in-graphicsmagick-for-node

**Comment by aqibmumtaz** [comment]
what does flop do?

Aqib

**Comment by DanMMX** [comment]
Flop mirrors the image horizontally, and not sure what you're trying to achieve, can you create an image with the expected result? With reversing the text the 2nd image in your stackoverflow question should be done

**Comment by aqibmumtaz** [comment]
as mentioned in question I m trying to print this text  "السلام عليكم" and output result should be same text
"السلام عليكم", but output text isn't being printed like that, all  I need to achieve is printing Arabic/Urdu text.

**Comment by DanMMX** [comment]
Which is the output when reversing the text? Isn't the desired either?

**Comment by aqibmumtaz** [comment]
No, output isn't desired even after reversing, this is reverse text "مكيلع مالسلا" of "السلام عليكم", If I try using reversed text it isn't producing required results as shown in stackoverflow question.

**Comment by DanMMX** [comment]
Maybe your problem would be easier to understand if you could provide a desired image, not sure if my browser encoding interferes but I see the reversed text pretty similar to the image you are receiving

**Comment by aqibmumtaz** [comment]
My INPUT text is "السلام عليكم" Matching Image:![screen shot 2015-08-26 at 10 34 46 am](https://cloud.githubusercontent.com/assets/1073190/9486359/97068bf6-4bdf-11e5-8d82-7a34e725ecc3.png)
 and Required OUTPUT text must be same as is "السلام عليكم" Matching Image:![screen shot 2015-08-26 at 10 34 58 am](https://cloud.githubusercontent.com/assets/1073190/9486814/78f06c26-4be5-11e5-88cf-48f9ce4d8d93.png)


Now Issue is output text doesn't matching with input text at all, Current Output text is:![1](https://cloud.githubusercontent.com/assets/1073190/9486830/8d644696-4be5-11e5-963b-287adc5beec8.png)


2ndly, If I tried reversing INPUT text as you suggested and used Reversed INPUT text "مكيلع مالسلا" Matching Image:![screen shot 2015-08-26 at 10 34 52 am](https://cloud.githubusercontent.com/assets/1073190/9486845/b23a6b30-4be5-11e5-9f19-243128699baa.png)
Still output text is as:![2](https://cloud.githubusercontent.com/assets/1073190/9486856/d22c1420-4be5-11e5-88da-e0f70a084600.png)

so it doesn't matter if I use reversed string or not, my output results are not desired.

is that help?

**Comment by aqibmumtaz** [comment]
@DanMMX is this question clear to you now?

**Comment by DanMMX** [comment]
oh ok, now that you put those images (for everything) is much clearer, I thought it was a matter of font type, now I see it's not. Sorry I don't think I know how to help you, if you come up with a solution let us know

**Comment by aqibmumtaz** [comment]
I think its gm support team who can come up with solution, you need to fix the way gm is handling Arabic  encoding, fonts and printing.

**Comment by m7md10** [comment]
It's old issue , but I was looking for a solution and I found this module
https://github.com/louy/Javascript-Arabic-Reshaper

`var ArabicReshaper = require('arabic-reshaper');
var txt = ArabicReshaper.convertArabic('العربية').split("").reverse().join("");`
then you can print it correctly with
` .drawText(60, 500,txt)`

hope someone find this useful .
